### PR TITLE
core/transaction: drop job that has unfulfilled required (Requires, BindsTo) dependencies

### PR DESCRIPTION
**Pull request by msekletar**
Resolves: #11338

### CODE REVIEW

File: src/core/transaction.c [hunk 1]
```diff
@@ -1000,8 +1000,14 @@ int transaction_add_job_and_dependencies(
                         UNIT_FOREACH_DEPENDENCY(dep, ret->unit, UNIT_ATOM_PULL_IN_START) {
                                 r = transaction_add_job_and_dependencies(tr, JOB_START, dep, ret, true, false, false, ignore_order, e);
                                 if (r < 0) {
-                                        if (r != -EBADR) /* job type not applicable */
+                                        if (r != -EBADR) { /* job type not applicable */
+                                                log_unit_error_errno(unit, r, "Cannot add dependency job: %s", bus_error_message(e, r));
```
Comment (poettering) [inline src/core/transaction.c]: adding the error logging here seems wrong. so far this function let's the caller log about errors it propagates, doesn't log on its own about them. It does log about some errors it eats up though. please stick to this rule.

**Comment by poettering** [comment]
This PR does make sense to me. But could you please add similar handling to the other dependency atoms we iterate through? i.e. whenever we fail the call we should really remove the job from the transaction again i guess. probably means adding a goto target to the bottom somewhere, that cleans this up, and then jump to it from all relevant places.

**Comment by msekletar** [comment]
Btw, regarding logging...I think at least debug message would be warranted.

**Comment by poettering** [comment]
> Btw, regarding logging...I think at least debug message would be warranted.

debug logging is always ok

**Comment by mrc0mmand** [comment]
> Btw, regarding logging...I think at least debug message would be warranted.

Definitely support at least a debug message, otherwise I'd have a hard time figuring out https://github.com/systemd/systemd/pull/27865 that was uncovered by this PR.

File: src/core/transaction.c [hunk 2]
```diff
@@ -1025,7 +1025,7 @@ int transaction_add_job_and_dependencies(
                                 r = transaction_add_job_and_dependencies(tr, JOB_VERIFY_ACTIVE, dep, ret, true, false, false, ignore_order, e);
                                 if (r < 0) {
                                         if (r != -EBADR) /* job type not applicable */
-                                                return r;
+                                                goto fail;
```
Comment (poettering) [inline src/core/transaction.c:1028]: hmm, for the UNIT_ATOM_PULL_IN_STOP case below I think we should add the same, no?

can you provide a rationale why not?

**Comment by poettering** [comment]
lgtm, just wondering about the UNIT_ATOM_PULL_IN_STOP case.

**Comment by msekletar** [comment]
> lgtm, just wondering about the UNIT_ATOM_PULL_IN_STOP case.

You are right, we should do it also for `UNIT_ATOM_PULL_IN_STOP`. I've added it in that case too.

File: src/core/transaction.c [hunk 3]
```diff
@@ -1102,6 +1102,12 @@ int transaction_add_job_and_dependencies(
         }
 
         return 0;
+fail:
+        /* Recursive call failed to add required jobs so let's drop top level job as well. */
+        log_unit_debug_errno(unit, r, "Cannot add dependency job: %s", bus_error_message(e, r));
```
Comment (poettering) [inline src/core/transaction.c]: maybe reword to say what you are doing here as effect, i.e. something like `log_unit_debug_errno(…, "Unable to add dependency job to transaction, deleting job %s/%s again", unit->id job_type_to_string(type)));`

you get the idea

File: src/core/transaction.c [hunk 4]
```diff
@@ -1102,6 +1102,12 @@ int transaction_add_job_and_dependencies(
...
+        transaction_delete_job(tr, ret, false);
```
Comment (poettering) [inline src/core/transaction.c]: please prefix the `false` with `/* delete_dependencies= */` as we do these days for function parameters that otherwise are just too opaque to figure out what they might mean.

**Comment by msekletar** [comment]
@poettering thank you for the suggestions. PTAL.

**Comment by poettering** [comment]
excellent! thanks!
